src/core/hiwire.c

```diff
@@ -6,144 +6,147 @@
-//    unoccupied slot: [version (5 bits)][next free index (25 bits)]0
-//
-// If a slot is occupied, the multipurpose field of the slotInfo contains a 24
-// bit reference count and an IS_DEDUPLICATED bit.
-//
-//      occupied slot: [version (5 bits)][refcount (24 bits)][IS_DEDUPLICATED bit]1
-//
-// References used by JsProxies are deduplicated which makes allocating/freeing
-// them more expensive.
-
-
-#define VERSION_SHIFT 26 // 1 occupied bit, 25 bits of index/nextfree/refcount, then the version
-#define INDEX_MASK            0x03FFFFFE // mask for index/nextfree
-#define REFCOUNT_MASK         0x03FFFFFC // mask for refcount
-#define VERSION_OCCUPIED_MASK 0xFc000001 // mask for version and occupied bit
-#define VERSION_MASK          0xFc000000 // mask for version
-#define OCCUPIED_BIT 1                   // occupied bit mask
-#define DEDUPLICATED_BIT 2               // is it deduplicated? (used for JsRefs)
-#define REFCOUNT_INTERVAL 4              // The refcount starts after OCCUPIED_BIT and DEDUPLICATED_BIT
-#define NEW_INFO_FLAGS 5                 // REFCOUNT_INTERVAL | OCCUPIED_BIT
-
-// Check that the constants are internally consistent
-_Static_assert(INDEX_MASK == ((1 << VERSION_SHIFT) - 2), "Oops!");
-_Static_assert((REFCOUNT_MASK | DEDUPLICATED_BIT) == INDEX_MASK, "Oops!");
-_Static_assert(VERSION_OCCUPIED_MASK == (~INDEX_MASK), "Oops!");
-_Static_assert(VERSION_OCCUPIED_MASK == (VERSION_MASK | OCCUPIED_BIT), "Oops!");
-_Static_assert(NEW_INFO_FLAGS == (REFCOUNT_INTERVAL | OCCUPIED_BIT), "Oops");
-
-#define HEAP_REF_TO_INDEX(ref) (((ref) & INDEX_MASK) >> 1)
-#define HEAP_INFO_TO_NEXTFREE(info) HEAP_REF_TO_INDEX(info)
-
-// The ref is always odd so this is truthy if info is even (meaning unoccupied)
-// or info has a different version than ref. Masking removes the bits that form
-// the index in the reference and the refcount/next free index in the info.
-#define HEAP_REF_IS_OUT_OF_DATE(ref, info) \
-  (((ref) ^ (info)) & VERSION_OCCUPIED_MASK)
-
-#define HEAP_IS_REFCNT_ZERO(info) (!((info) & REFCOUNT_MASK))
-#define HEAP_IS_DEDUPLICATED(info) ((info) & DEDUPLICATED_BIT)
-
-#define HEAP_INCREF(info) info += REFCOUNT_INTERVAL
-#define HEAP_DECREF(info) info -= REFCOUNT_INTERVAL
-
-// increment the version in info.
-#define _NEXT_VERSION(info) (info + (1 << VERSION_SHIFT))
-// assemble version, field, and occupied
-#define _NEW_INFO(version, field_and_flag) \
-  (((version) & VERSION_MASK) | (field_and_flag))
-
-// make a new reference with the same version as info and the given index.
-#define HEAP_NEW_REF(index, info) _NEW_INFO(info, ((index) << 1) | 1)
-// new occupied info: same version as argument info, NEW_INFO_FLAGS says occupied with refcount 1
-#define HEAP_NEW_OCCUPIED_INFO(info) _NEW_INFO(info, NEW_INFO_FLAGS)
-// new unoccupied info, increment version and nextfree in the field
-#define FREE_LIST_INFO(info, nextfree) _NEW_INFO(_NEXT_VERSION(info), (nextfree) << 1)
+EM_JS(void, hiwire_invalid_id, (int type, JsRef ref), {
```

Comment:
What is this function for? Debugging?

Reply:
Yeah it's called if someone dereferences an id that is malformed, is freed, or hasn't been assigned yet. As a to-do I should add back in the much more detailed internal error handling we had before...

Reply:
Okay I added extra info here and a test.